usr/sbin/p11sak/p11sak.c

```diff
@@ -797,28 +797,81 @@ static CK_RV read_ec_args(const char *ECcurve, CK_ATTRIBUTE *pubattr,
 }
 /**
  * Builds two CKA_LABEL attributes from given label.
+ * By default the specified label is extended with ":pub" and ":prv" for the
+ * public and private key objects.
+ * To set 2 different labels for public and private keys, separate them by
+ * colon: "pub-label:prv-label".
+ * To set the exact same label for public and private key, use "pub-label:="
+ * To specify a colon or a equal character within a label, it must be escaped
+ * by a back slash:  * "abc\:xyz" results in "abc:xyz".
  */
 static CK_RV set_labelpair_attr(const char *label, CK_ATTRIBUTE *pubattr,
                                 CK_ULONG *pubcount, CK_ATTRIBUTE *prvattr,
                                 CK_ULONG *prvcount)
 {
-    char *publabel;
-    char *prvlabel;
+    char *publabel = NULL;
+    char *prvlabel = NULL;
+    unsigned int i;
 
-    if (!(publabel = malloc(strlen(label) + 5))) {
-        fprintf(stderr, "Error allocating space for publabel\n");
-        return CKR_HOST_MEMORY;
+    for (i = 0; i < strlen(label); i++) {
+        if (label[i] == ':') {
+            if (i > 0 && label[i - 1] == '\\') /* escaped colon */
+                continue;
+
+            if (!(publabel = strndup(label, i))) {
+                fprintf(stderr, "Error allocating space for publabel\n");
+                return CKR_HOST_MEMORY;
+            }
+            if (!(prvlabel = strdup(&label[i + 1]))) {
+                fprintf(stderr, "Error allocating space for prvlabel\n");
```

Comment:
Are you intentionally using strndup above, but strdup here?

Reply:
`strndup()` is for the first part of the string, `strdup()` for the rest of the string.

Reply:
Correct, the first part (up to the colon) is `i` characters long, thus a `strndup` with `i` characters only. The second part, starting after the colon is zero terminated, so a `strdup` is used.